torchsnapshot/storage_plugins/fsspec.py

```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/env python3
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+import asyncio
+import io
+import os
+
+import fsspec
+
+from torchsnapshot.io_types import ReadIO, StoragePlugin, WriteIO
+
+__all__ = ["FSSpecStoragePlugin"]
+
+
+class FSSpecStoragePlugin(StoragePlugin):
+    def __init__(self, root: str) -> None:
+        root_items = root.split("://")
+        if len(root_items) != 2:
+            raise ValueError("only protocol://path is supported by fsspec plugin")
+        protocol, self.root = root_items
+        if not protocol.startswith("fsspec-"):
+            raise ValueError(
+                f"Invalid protocol: {protocol}, Only fsspec-* protocols are supported"
+            )
+        self.fs = fsspec.filesystem(protocol=protocol.removeprefix("fsspec-"))
+        self._session = None
+        self._lock = asyncio.Lock()
+
+    async def _init_session(self) -> None:
+        async with self._lock:
+            if self._session is None:
+                self._session = await self.fs.set_session()
+
+    async def write(self, write_io: WriteIO) -> None:
+        await self._init_session()
+        path = os.path.join(self.root, write_io.path)
+        await self.fs._pipe_file(path, bytes(write_io.buf))
```

Comment:
`write_io.buf` can be a `memoryview`, in which case `bytes(write_io.buf)` will result in an extra copy. Do you know if `_pipe_file` accepts `memoryview`s?

Reply:
actuallly, when I test my implement with the fsspec's s3fs plugin, it will result to error if we just put a memoryview as a _pipe_file input. Because it seems that the memoryview don't has the len method.

the implement of s3fs plugin is here:
https://github.com/fsspec/s3fs/blob/main/s3fs/core.py#L998

Reply:
Never mind. After my double checking of using memoryview, I get the exception from botocore.validate:

    def serialize_to_request(self, parameters, operation_model):
        input_shape = operation_model.input_shape
        if input_shape is not None:
            report = self._param_validator.validate(
                parameters, operation_model.input_shape
            )
            if report.has_errors():
>               raise ParamValidationError(report=report.generate_report())
E               botocore.exceptions.ParamValidationError: Parameter validation failed:
E               Invalid type for parameter Body, value: <memory at 0x7fbfa202f280>, type: <class 'memoryview'>, valid types: <class 'bytes'>, <class 'bytearray'>, file-like object

../../../.conda/envs/torchsnapshot/lib/python3.10/site-packages/botocore/validate.py:381: ParamValidationError